File: 991.broken-calculator.cpp
/*
 * @lc app=leetcode id=991 lang=cpp
 *
 * [991] Broken Calculator
 */
#include <unordered_map>
#include <climits>
using namespace std;
// @lc code=start
// t:O(log(target)), s:O(1)
class Solution
{
    int start_value;

public:
    int brokenCalc(int startValue, int target)
    {
        start_value = startValue;
        return dfs(target);
    }
    int dfs(int n)
    {
        if (n == start_value)
        {
            return 0;
        }
        return n % 2 == 0 && n > start_value ? 1 + dfs(n / 2) : n < start_value ? start_value - n
                                                                                : 1 + dfs(n + 1);
    }
};
// @lc code=end

File: py/991.broken-calculator.py
#
# @lc app=leetcode id=991 lang=python3
#
# [991] Broken Calculator
#

# @lc code=start
class Solution:
    def brokenCalc(self, startValue: int, target: int) -> int:
        def dfs(n: int) -> int:
            if n == startValue:
                return 0
            if n % 2 == 0 and n > startValue:
                return 1 + dfs(n//2)
            elif n < startValue:
                return startValue-n
            else:
                return 1 + dfs(n+1)
        return dfs(target)
# @lc code=end
